casper/src/main/scala/coop/rchain/casper/safety/CliqueOracle.scala

```diff
@@ -0,0 +1,214 @@
+package coop.rchain.casper.safety
+
+import cats.Show
+import cats.data.OptionT
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import coop.rchain.casper.util.{Clique, ProtoUtil}
+import coop.rchain.dag.{DagOps, DagReader}
+import coop.rchain.metrics.Metrics.Source
+import coop.rchain.metrics.implicits._
+import coop.rchain.metrics.{Metrics, Span}
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+
+object CliqueOracle {
+
+  final case class Justification[M, V](message: M, creator: V)
+
+  // map of validator id to stake bonded
+  type WeightMap[V] = Map[V, Long]
+
+  implicit val baseMetricsSource: Source = coop.rchain.casper.CasperMetricsSource
+
+  private def computeMaxCliqueWeight[F[_]: Concurrent: Span, M, V](
+      target: M,
+      dag: DagReader[F, M],
+      agreeingWeightMap: WeightMap[V],
+      agreeingLatestMessages: Map[V, M],
+      toJustificationF: M => F[Justification[M, V]],
+      justificationsF: M => F[Set[Justification[M, V]]],
+      heightF: M => F[Long]
+  )(implicit show: Show[M]): F[Long] = Span[F].traceI("compute-max-clique-weight") {
```

Comment:
`Show` can be defined as context bound parameter instead of "normal" implicit parameter similar to `F`.
```scala
def computeMaxCliqueWeight[F[_]: Concurrent: Span, M: Show, V]
```